### PR TITLE
Friendlier service decorators for ScriptService

### PR DESCRIPTION
**Pull request by rhysparry**
# Background

Integration tests for service decorators extensively use `HookServiceMethod` to apply decorators to different ScriptService implementations.

Using the method this way has a few disadvantages:

- It was designed to apply decorators to methods that take a single parameter and return an async result. None of the tests leveraged this functionality, which led to lambda method signatures like `(_, _)`, which could lead to confusion.
- It requires passing in the `TentacleConfigurationTestCase`, which is not used by these decorators.
- It matched the decorated method by name, resulting in usages like `nameof(IScriptServiceV2Async.StartScriptAsync)`, leading to confusion as the decorator was applied to all script services.

Adding a builder to manage the addition of decorators to each service makes the intention more explicit, and the additional noise of the previous technique is resolved.

# Results

- Adds the new decorators and ensures `After` decorators are implemented for the 3 script services.
- Updates tests that leveraged `HookServiceMethod` to decorate script services to switch to the new `DecorateAllScriptServicesWith` method.
- [SC-69397]

## Before

Initially, code would need to map by name:

```csharp
await using var clientAndTentacle = await tentacleConfigurationTestCase.CreateBuilder()
    .WithTentacleServiceDecorator(new TentacleServiceDecoratorBuilder()
        .HookServiceMethod(
            tentacleConfigurationTestCase,
            nameof(IAsyncClientScriptServiceV2.GetStatusAsync),
            async (_, _) => await beforeGetStatus(),
            async (_, _) => await afterGetStatus())
        .HookServiceMethod(
            tentacleConfigurationTestCase,
            nameof(IAsyncClientScriptServiceV2.CancelScriptAsync),
            async (_, _) => await beforeCancelScript())
    .Build(CancellationToken);
```

## After

Explicit method decorators are used, with the intention of each decorator clear:

```csharp
await using var clientAndTentacle = await tentacleConfigurationTestCase.CreateBuilder()
    .WithTentacleServiceDecorator(new TentacleServiceDecoratorBuilder()
        .DecorateAllScriptServicesWith(u => u
            .BeforeGetStatus(
                async () => await beforeGetStatus())
            .AfterGetStatus(
                async () => await afterGetStatus())
            .BeforeCancelScript(
                async (_, _) => await beforeCancelScript()))
    .Build(CancellationToken);
```

# How to review this PR

Quality :heavy_check_mark:

# Pre-requisites

- [x] I have read [How we use GitHub Issues](https://github.com/OctopusDeploy/Issues/blob/main/docs/CONTRIBUTING.internal.md) for help deciding when and where it's appropriate to make an issue.
- [x] I have considered informing or consulting the right people, according to the [ownership map](https://whimsical.com/ownership-map-NzbiD4HJyvhC9jNJNfS6TG).
- [x] I have considered appropriate testing for my change.

### CODE REVIEW

File: source/Octopus.Tentacle.Tests.Integration/Util/Builders/Decorators/ScriptServiceDecoratorBuilder.cs [hunk 1]
```diff
@@ -13,20 +14,37 @@ public class ScriptServiceDecoratorBuilder
...
-        private GetStatusClientDecorator getStatusFunc = async (inner, command, options) => await inner.GetStatusAsync(command, options);
-        private CancelScriptClientDecorator cancelScriptFunc = async (inner, command, options) => await inner.CancelScriptAsync(command, options);
-        private CompleteScriptClientDecorator completeScriptAction = async (inner, command, options) => await inner.CompleteScriptAsync(command, options);
+        StartScriptClientDecorator startScriptFunc = async (inner, command, options) => await inner.StartScriptAsync(command, options);
+        GetStatusClientDecorator getStatusFunc = async (inner, command, options) => await inner.GetStatusAsync(command, options);
+        CancelScriptClientDecorator cancelScriptFunc = async (inner, command, options) => await inner.CancelScriptAsync(command, options);
+        CompleteScriptClientDecorator completeScriptAction = async (inner, command, options) => await inner.CompleteScriptAsync(command, options);
 
-        public ScriptServiceDecoratorBuilder BeforeStartScript(Func<Task> beforeGetStatus)
+        public ScriptServiceDecoratorBuilder BeforeStartScript(Func<Task> beforeStartScript)
```
Comment (rhysparry) [inline source/Octopus.Tentacle.Tests.Integration/Util/Builders/Decorators/ScriptServiceDecoratorBuilder.cs:22]: Renamed parameter for clarity

File: source/Octopus.Tentacle.Tests.Integration/Util/Builders/Decorators/ScriptServiceDecoratorBuilder.cs [hunk 2]
```diff
@@ -13,20 +14,37 @@ public class ScriptServiceDecoratorBuilder
...
         {
-            return DecorateStartScriptWith(async (inner, scriptStatusRequest, options) =>
+            return DecorateStartScriptWith(async (inner, startScriptCommand, options) =>
```
Comment (rhysparry) [inline source/Octopus.Tentacle.Tests.Integration/Util/Builders/Decorators/ScriptServiceDecoratorBuilder.cs:24]: Another renamed parameter

File: source/Octopus.Tentacle.Tests.Integration/Util/Builders/Decorators/ScriptServiceV2DecoratorBuilder.cs [hunk 3]
```diff
@@ -25,10 +25,32 @@ public ScriptServiceV2DecoratorBuilder BeforeStartScript(Func<Task> beforeStartS
 
         public ScriptServiceV2DecoratorBuilder BeforeStartScript(Func<IAsyncClientScriptServiceV2, StartScriptCommandV2, HalibutProxyRequestOptions, Task> beforeStartScript)
         {
-            return DecorateStartScriptWith(async (inner, scriptStatusRequestV2, options) =>
+            return DecorateStartScriptWith(async (inner, startScriptCommandV2, options) =>
```
Comment (rhysparry) [inline source/Octopus.Tentacle.Tests.Integration/Util/Builders/Decorators/ScriptServiceV2DecoratorBuilder.cs:28]: Renamed parameter for clarity

**Review by rhysparry — COMMENTED** [review]
I recommend hiding whitespace changes when reviewing the tests to avoid noise from changes in the indent level

image

File: source/Octopus.Tentacle.Tests.Integration/Support/ExceptionContractAssertionBuilder.cs [hunk 4]
```diff
@@ -32,6 +32,7 @@ public ExceptionContract Build()
                             $"An error occurred when sending a request to '{clientAndTentacle.ServiceEndPoint}/', before the request could begin:  Received an unexpected EOF or 0 bytes from the transport stream",
                             $"An error occurred when sending a request to '{clientAndTentacle.ServiceEndPoint}/', before the request could begin: Unable to write data to the transport connection: Broken pipe",
                             $"An error occurred when sending a request to '{clientAndTentacle.ServiceEndPoint}/', before the request could begin: Unable to read data from the transport connection: Broken pipe",
+                            $"An error occurred when sending a request to '{clientAndTentacle.ServiceEndPoint}/', before the request could begin: Unable to read data from the transport connection: Connection reset by peer.",
```
Comment (LukeButters) [inline source/Octopus.Tentacle.Tests.Integration/Support/ExceptionContractAssertionBuilder.cs:35]: Did we miss one?

**Review by LukeButters — APPROVED** [review]
LGTM - the test read waaaay better now thanks

**Pull request merged**